objects/src/transaction/tx_witness.rs

```diff
@@ -0,0 +1,147 @@
+    consumed_notes_hash: Digest,
+    tx_script_root: Option<Digest>,
+    program: Program,
+    advice_witness: AdviceInputs,
+}
+
+impl TransactionWitness {
+    /// Creates a new [TransactionWitness] from the provided data.
+    pub fn new(
+        account_id: AccountId,
+        initial_account_hash: Digest,
+        block_hash: Digest,
+        consumed_notes_hash: Digest,
+        tx_script_root: Option<Digest>,
+        program: Program,
+        advice_witness: AdviceInputs,
+    ) -> Self {
+        Self {
+            account_id,
+            initial_account_hash,
+            block_hash,
+            consumed_notes_hash,
+            tx_script_root,
+            program,
+            advice_witness,
+        }
+    }
+
+    /// Returns the account id of the account the transaction is executed against.
+    pub fn account_id(&self) -> &AccountId {
+        &self.account_id
+    }
+
+    /// Returns the initial account hash of the account the transaction is executed against.
+    pub fn initial_account_hash(&self) -> &Digest {
+        &self.initial_account_hash
+    }
+
+    /// Returns the block hash of the latest known block.
+    pub fn block_hash(&self) -> &Digest {
+        &self.block_hash
+    }
+
+    /// Returns the consumed notes hash.
+    pub fn consumed_notes_hash(&self) -> &Digest {
+        &self.consumed_notes_hash
+    }
+
+    /// Returns a vector of [ConsumedNoteInfo] for all of the consumed notes in the transaction.
+    ///
+    /// # Errors
+    /// - If the consumed notes data is not found in the advice map.
+    /// - If the consumed notes data is not well formed.
+    pub fn consumed_notes_info(&self) -> Result<Vec<ConsumedNoteInfo>, TransactionWitnessError> {
+        // fetch consumed notes data from the advice map
+        let notes_data = self
+            .advice_witness
+            .mapped_values(&self.consumed_notes_hash.as_bytes())
+            .ok_or(TransactionWitnessError::ConsumedNoteDataNotFound)?;
+
+        // extract the notes from the first fetch and instantiate a vector to hold
+        // [ConsumedNoteInfo].
+        let num_notes = notes_data[0].as_int();
+        let mut consumed_notes_info = Vec::with_capacity(num_notes as usize);
+
+        // iterate over the notes and extract the nullifier and script root
+        let mut note_ptr = 1;
+        while note_ptr < notes_data.len() {
+            // make sure there is enough data to read (note data is well formed)
+            if note_ptr + 5 * WORD_SIZE > notes_data.len() {
+                return Err(TransactionWitnessError::InvalidConsumedNoteDataLength);
+            }
+
+            // compute the nullifier and script root
+            let nullifier =
+                Hasher::hash_elements(&notes_data[note_ptr..note_ptr + (4 * WORD_SIZE)]);
+            let script_root: Word = notes_data[note_ptr + WORD_SIZE..note_ptr + (2 * WORD_SIZE)]
+                .try_into()
+                .expect("word is well formed");
```

Comment:
nit: I would probably create a small helper function to extract this and would pass `&note_data[note_ptr..]` into it. This way, understanding how the elements are extracted from the note data becomes a bit easier.